commit ff91a417f3ec2562a2fe0511185332ff4a180a4a
Author: Marcus Cooper <codekipper@gmail.com>
Date:   Sun Nov 25 09:48:57 2018 +0100

    spdif for kodi

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pine64.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-pine64.dts
index 467494745a21..f7cbe8514581 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pine64.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pine64.dts
@@ -261,7 +261,7 @@
 
 /* On Euler connector */
 &spdif {
-	status = "disabled";
+	status = "okay";
 };
 
 /* On Exp and Euler connectors */
